### ISSUE TITLE
Fragment Inheritance not Pushed to Supertype

### ISSUE THREAD

**Issue by chriszwickerocteris** [eclipse/xtext#1527]
Given the following models

```
Model:
	foobars+=FooBar*;
	
FooBar :
  Foo | Bar
;

Foo :
  "foo" name=ID
  FooBarBody
;

Bar :
  "bar" name=ID
  FooBarBody
;

fragment FooBarBody :
  content = STRING
;
```

we get the following generated ecore model

```
dsl
+- Model
   +- foobars: Foobar
+- Foobar
+- Foo -> FooBar, FooBarBody
   +- name: EString
+- Bar -> FooBar, FooBarBody
   +- name: EString
+- FooBarBody
   +- content: EString
```

What I would expect, though, would be the following

```
dsl
+- Model
   +- foobars: Foobar
+- Foobar -> FooBarBody
+- Foo -> FooBar
   +- name: EString
+- Bar -> FooBar
   +- name: EString
+- FooBarBody
   +- content: EString
```

Only if this is given can we really avoid code duplication (which is the problem I understand fragments to solve), otherwise in generator code we have to add casts where we don't expect to need them...

Of course as a workaround, we can use imported metamodels, but it would be nice if these were generated as expected.

**Comment by cdietrich** [comment]
@szarnekow is this intended or a bug? should the fragment be moved to the parent?

**Comment by szarnekow** [comment]
From the top of my head, I'd say the current behaviour is "works as designed". Also it's not easy to judge what the actual expectation for this is.
Assuming we wanted to pull the super type up, the inference would completely change as soon as a third alternative is added to the rule FooBar.
Moving the call to the fragment rule up in the grammar call graph seems to be a good solution for this particular case.

**Comment by chriszwickerocteris** [comment]
Would it be wrong to assume the expected behaviour would be what would happen if we still didn't have fragments and the code now extracted into a fragment was duplicated into each type now calling the fragment rule? Or am I mistaken in assuming the fragment feature was introduced precisely to avoid duplication?

**Comment by szarnekow** [comment]
Yes and no. The exact equivalent would be `fragment FooBarBody returns FooBar: ..` but with `fragment FooBarBody` you introduce an additional type to sport multiple inheritance / aspects of the model. 
In that sense, yes:
```
FooBar : Foo | Bar;
Foo: name = ID content = STRING;
Bar: name = ID content = STRING;
```
will move both attributes `name` and `content` into `FooBar`.
```
FooBar : Foo | Bar | Zonk;
Foo: name = ID content = STRING;
Bar: name = ID content = STRING;
Zonk: name =ID;
```
Would force `content` to be defined in `Foo` and `Bar` only.
With fragments being part of the game:
```
FooBar: Foo | Bar;
Foo: name = ID FooBarBody;
Bar: name = ID FooBarBody;
fragment FooBarBody: content = STRING;
```
Iff we made FooBarBody a supertype of FooBar, we could even argue that name should be pushed into that supertype. What makes name different from content for this case? On the other hand, it could be a deliberate decision to introduce multiple inheritance, e.g. FooBarBody being a second super type of Foo and Bar. That would be impossible if we always pushed the type up.

Long story short: Mere duplication can be avoided by an explicit return type on the fragment: 
`fragment FooBarBody returns FooBar: ..`
or by using wildcard fragments:
`fragment FooBarBody *: ..`
Both would suppress the additional type FooBarBody though.

**Comment by chriszwickerocteris** [comment]
Thanks, didn't realize there were wildcard fragments (as far as I could see, the feature hasn't yet made it into the documentation, or did I miss something?). For my use case, I can easily do without the extra supertype.

Just as a sidenote: intuitively I would have assumed the keyword "fragment" to already behave like "wildcard fragment" and to require an explicit `returns FooBarBody` in order to actually create a type. What would the typical user base their interpretation of the word "fragment" on? Eclipse fragments? Something else?

**Comment by chriszwickerocteris** [comment]
And in response to your example, I think with `Zonk` in play, fragment could behave as is, creating a separate type which is inherited by `Foo` and `Bar`. This doesn't seem to be inconsistent with what would happen without fragments (assuming you don't use `FooBarBody` explicitly, everything would be as it was without fragments, no?

**Comment by cdietrich** [comment]
ping @szarnekow is there anything todo on Xtext side?

**Comment by szarnekow** [comment]
Generally speaking I'd like to define and seal the semantics of fragments to avoid future confusion. Not sure if we find the time in the foreseeable future though.
